File: backend/app/models/challenge.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class ChallengeResponse(BaseModel):
    """A style challenge (CLAUDE.md §1 pillar 4, §24)."""

    id: str
    slug: str
    title: str
    prompt: str | None = None
    cover_url: str | None = None
    starts_at: datetime
    ends_at: datetime | None = None
    entry_count: int = 0
    joined_by_me: bool = False


class ChallengeJoin(BaseModel):
    """Enter a challenge by linking one of your own OOTD posts to it."""

    post_id: UUID


class ChallengeEntryResponse(BaseModel):
    """A post entered into a challenge, with its author (for the entries feed)."""

    id: str
    challenge_id: str
    post_id: str
    user_id: str
    author_name: str | None = None
    image_url: str | None = None
    caption: str | None = None
    created_at: datetime = Field(...)

File: app/lib/data/models/challenge.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'challenge.freezed.dart';
part 'challenge.g.dart';

/// A style challenge (CLAUDE.md §1 pillar 4). JSON keys match the
/// `/v1/challenges` response (the challenge + entry count + the viewer's
/// entered state).
@freezed
abstract class Challenge with _$Challenge {
  const factory Challenge({
    required String id,
    required String slug,
    required String title,
    String? prompt,
    @JsonKey(name: 'cover_url') String? coverUrl,
    @JsonKey(name: 'starts_at') required DateTime startsAt,
    @JsonKey(name: 'ends_at') DateTime? endsAt,
    @JsonKey(name: 'entry_count') @Default(0) int entryCount,
    @JsonKey(name: 'joined_by_me') @Default(false) bool joinedByMe,
  }) = _Challenge;

  factory Challenge.fromJson(Map<String, dynamic> json) =>
      _$ChallengeFromJson(json);
}
